### ISSUE TITLE
Tensorflow Lite C API + GPU delegate crash on exit

### ISSUE THREAD

**Issue by DwayneDuane** [tensorflow/tensorflow#51751]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): master/nightly
- Python version: N/A
- Bazel version (if compiling from source): 3.7.2
- GCC/Compiler version (if compiling from source): gcc 8.4.0
- CUDA/cuDNN version: cuda 11.1
- GPU model and memory: 4GB VRAM, 8GM system RAM

**Describe the current behavior**
Using the C API of tensorflow lite, with GPU delegate (GPU = NVidia GTX 1650), I get a crash on exit. This only happens when I build the latest mater or nightly branch. **The v2.6.0 tag does not have this problem.** 
Here is my code to reproduce the issue:
```
#include "tensorflow/lite/c/c_api.h"
#include "tensorflow/lite/delegates/gpu/delegate.h"

int main(void)
{
    TfLiteGpuDelegateOptionsV2 opts = TfLiteGpuDelegateOptionsV2Default();

    TfLiteDelegate* gpuDelegate = TfLiteGpuDelegateV2Create(&opts);
    TfLiteInterpreterOptions* options = TfLiteInterpreterOptionsCreate();

    TfLiteInterpreterOptionsAddDelegate(options, gpuDelegate);

    //This problem should be be reproducible using any convolutional model
    TfLiteModel* model = TfLiteModelCreateFromFile("../mymodel.tflite");
    TfLiteInterpreter* interpreter = TfLiteInterpreterCreate(model, options);

    TfLiteInterpreterAllocateTensors(interpreter);

    TfLiteInterpreterInvoke(interpreter);

    TfLiteInterpreterDelete(interpreter);
    TfLiteGpuDelegateV2Delete(gpuDelegate);
    TfLiteInterpreterOptionsDelete(options);
    TfLiteModelDelete(model);
}
```

The stacktrace I get from gdb is:
```
INFO: Initialized OpenCL-based API.
INFO: Created 1 GPU delegate kernels.

Thread 1 "dlib_test" received signal SIGSEGV, Segmentation fault.
0x00007fffed25d1a0 in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
(gdb) bt
#0  0x00007fffed25d1a0 in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
#1  0x00007fffed15655c in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
#2  0x00007fffed153b08 in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
#3  0x00007fffed155c22 in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
#4  0x00007ffff7531a06 in tflite::gpu::cl::Buffer::Release() ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so
#5  0x00007ffff752fadf in tflite::gpu::cl::InferenceContext::~InferenceContext() ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so
#6  0x00007ffff7529369 in tflite::gpu::cl::(anonymous namespace)::InferenceRunnerImpl::~InferenceRunnerImpl() ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so
#7  0x00007ffff75222e2 in tflite::gpu::(anonymous namespace)::DelegatePrepare(TfLiteContext*, TfLiteDelegate*)::{lambda(TfLiteContext*, void*)#2}::_FUN(TfLiteContext*, void*) ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlit---Type <return> to continue, or q <return> to quit---
e_c.so
#8  0x00007ffff751c654 in tflite::Subgraph::CleanupNode(int) ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so
#9  0x00007ffff751c6bf in tflite::Subgraph::~Subgraph() ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so
#10 0x00007ffff751c879 in tflite::Subgraph::~Subgraph() ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so
#11 0x00007ffff7a00697 in tflite::Interpreter::~Interpreter() ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so
#12 0x00007ffff7516ef1 in TfLiteInterpreterDelete ()
   from /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so
#13 0x0000555555554bf8 in main ()
```
When I run my program using valgrind, I get:
```
==19931== Conditional jump or move depends on uninitialised value(s)
==19931==    at 0x4FCAA64: tflite::gpu::ConvPowerVR::GenerateConv[abi:cxx11](tflite::gpu::GpuInfo const&, tflite::gpu::OperationDef const&, bool, tflite::gpu::ConvPowerVR::ConvParams const&)::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#3}::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const [clone .isra.283] (in /home/<user>/Desktop/tensorflow_lite_c_master/lib/x86_64/libtensorflowlite_c.so)
```

**Comment by vulkomilev** [comment]
I am not sure but it looks like that this issue may come from not enough memory . At least when I not provide enough memory to my tensorflow program I get funny errors like illegal memory access or uninitialized layer .Can you confirm that you have enough memory?

**Comment by abattery** [comment]
@impjdi could you take a look at this?

**Comment by DwayneDuane** [comment]
@vulkomilev I have a fairly powerful machine with 8GB of system ram and an NVidia GTX 1650 with 4GB vram. I was able to repro the issue using the code I showed above and a NasnetMobile (FP32) net I download from https://www.tensorflow.org/lite/guide/hosted_models . This is a relatively small model and shouldn't cause memory pressure. Also, as I mentioned in the original post, I don't get this crash when building tflite from v2.6.0 tag. So this feels more like a bug that was recently introduced.
